geth/node/node_test.go

```diff
@@ -0,0 +1,96 @@
+package node
+
+import (
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestInboundConnections(t *testing.T) {
+	type testCase struct {
+		name    string
+		max     int
+		percent int
+	}
+
+	for _, tc := range []testCase{
+		{
+			name:    "All",
+			max:     4,
+			percent: 100,
+		},
+		{
+			name:    "Half",
+			max:     4,
+			percent: 50,
+		},
+		{
+			name:    "FloorDivision",
+			max:     4,
+			percent: 70,
+		},
+		{
+			name:    "None",
+			max:     4,
+			percent: 0,
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			key, _ := crypto.GenerateKey()
+			main := p2p.Server{
+				Config: p2p.Config{
+					NoDiscovery:    true,
+					PrivateKey:     key,
+					InboundPercent: tc.percent,
+					MaxPeers:       tc.max,
+					ListenAddr:     "127.0.0.1:0",
+				},
+			}
+			require.NoError(t, main.Start())
+			defer main.Stop()
+
+			rst := make(chan *p2p.PeerEvent, main.Config.MaxPeers)
+			sub := main.SubscribeEvents(rst)
+			defer sub.Unsubscribe()
+
+			peers := make([]*p2p.Server, 0, main.Config.MaxPeers)
+			for i := 0; i < main.Config.MaxPeers; i++ {
+				key, _ := crypto.GenerateKey()
+				peer := &p2p.Server{
+					Config: p2p.Config{
+						NoDiscovery: true,
+						PrivateKey:  key,
+						MaxPeers:    1,
+						ListenAddr:  "127.0.0.1:0",
+					},
+				}
+				assert.NoError(t, peer.Start())
```

Comment:
Not sure if this `assert` makes sense. If the p2p server fails to start, it's a problem with configuration or system resources. What's the advantage of continuing with the rest of the test?